File: services/eval/src/generator/jsonl-writer.ts
import { mkdir, rename, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import type { GoldenRecord } from "./types.js";

/**
 * Serialize records to JSONL and write atomically.
 *
 * WHY write-then-rename: a half-finished `.jsonl` is worse than no file at
 * all, because downstream eval runs would silently load a partial dataset.
 * `rename` within the same filesystem is atomic on POSIX, so readers either
 * see the previous file or the complete new one.
 */
export async function writeJsonl(path: string, records: readonly GoldenRecord[]): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  const tempPath = `${path}.tmp-${process.pid}-${Date.now()}`;
  const body = records.map((r) => JSON.stringify(r)).join("\n") + (records.length > 0 ? "\n" : "");
  await writeFile(tempPath, body, "utf8");
  await rename(tempPath, path);
}

File: services/eval/src/rag_eval/generator/jsonl_writer.py
import json
import os
import time
from collections.abc import Sequence
from pathlib import Path

from rag_eval.generator.models import GoldenRecord


def write_jsonl(path: str | Path, records: Sequence[GoldenRecord]) -> None:
    """Serialize records to JSONL and write atomically.

    Write-then-rename because a half-finished ``.jsonl`` is worse than no file
    at all — downstream eval runs would silently load a partial dataset.
    ``os.replace`` within the same directory is atomic on POSIX, so readers
    either see the previous file or the complete new one.
    """
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    temp_path = target.with_name(f"{target.name}.tmp-{os.getpid()}-{time.time_ns()}")

    lines = [json.dumps(record.to_json_dict(), ensure_ascii=False) for record in records]
    body = "\n".join(lines) + ("\n" if records else "")

    temp_path.write_text(body, encoding="utf-8")
    os.replace(temp_path, target)
